### PR TITLE
Fix #1118 and other issues

### LINKED ISSUE #1118: Search only matches on exact strings
### Steps to reproduce the error

Open any project, search for any partial term. Example: Project "Chocolate Factory" contains a party named David.

### Actual behavior

Search returns zero results if there is no match for the exact string entered. Example: Searching "Chocolate Factory" for davi returns zero results.

### Expected behavior

According to the requirements in #826, an exact string match should be conducted iff the search term is enclosed in quotation marks. A string entered without quotation marks should return the results that best match. Example: Searching for davi without quotes should return the party named David.

### LINKED ISSUE #1354: Archived resources are also shown in search results.
### Steps to reproduce the error

1. Create a resource and archive it.
2. Use search to find the resource.
3. Clicking on the resource from the search results will give DoesNotExist exception.

### Actual behavior

Archived resources appear in search results.

### Expected behavior

Archived resources should not appear in project search results.

### PR DESCRIPTION
**Pull request by seav**
### Proposed changes in this pull request
- Use default "or":
    - Make search use "or" by default and allow users to specify that terms are required or excluded by prefixing them with `+` and `-` respectively.
    - Implement this by using the pyparsing Python library. See the new **search/parser.py** source file and the `parse_query()` function for details. Before, the platform just passes the UI query string directly to ES which does the parsing. Now, we can control our search query syntax by having the platform do the parsing and then constructing the ES search DSL based on the parsed results.
    - Terms without the prefixed `+` or `-` are mapped to the ES `should` boolean clause, terms with the prefixed `+` are mapped to the ES `must` boolean clause, and terms with the prefixed `-` are mapped to the ES `must_not` boolean clause. [[ES reference](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html)]
    - In addition, full-text search is now limited to ES fields that we can reasonably expect users to want to search on.
    - Update the search guidelines on the search HTML page.
- Fix #1118 (search only matches on exact strings):
    - Make search fuzzy by default and allow users to specify exact matching on whole phrases by using quotation marks. However, make the "not" operator (`-`) use exact matching (we don't want to accidentally exclude things that fuzzily match the "not" term).
    - Implement the fuzziness by adding a second fuzzy matching clause to `should` and `must` clauses. The fuzziness (aka [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance)) is set to 1 for terms that are 2 to 3 characters long and to 2 for longer terms. If the term is only 1 character long, no fuzzy matching clause is added.
    - Implement exact matching by using pyparsing's `QuotedString` object. These are then mapped to the ES `phrase` match type and no fuzzy matching clauses are added as mentioned in the previous bullet. 
    - Update the search guidelines on the search HTML page.
- Fix #1354 (archived resources are included in results):
    - Add the following additional `must_not` clause to the ES query DSL: `{"match": {"archived": true}}`
- Fix remaining DataTables Ajax warnings:
    - Before I assumed that connecting to the ES API will always succeed and then flagging an error to the user only if the HTTP response code is not 200. Now HTTP connection errors to ES are also caught (by catching `requests.exceptions.RequestException` exceptions). This should now avoid the JavaScript "DataTables warning" popup that appears when there is a connection error.
    - Also add an explicit `timeout=10` parameter to HTTP requests. Without this, requests will wait indefinitely. [[Requests reference](http://docs.python-requests.org/en/master/user/advanced/#timeouts)]
- Fix the issue that the search Ajax endpoint is hit twice for every search:
    - Remove a redundant call to the `draw()` DataTables API function.
- Other improvements:
    - Limit the ES types to be searched to just `spatial`, `party`, and `resource` by specifying them in the ES API URL. Before, no ES type was specified so this defaults to searching all ES types, which also includes the `project` ES type.
        - Adjust the Mock ES implementation to handle this change of ES API URL.

### When should this PR be merged
- Within Sprint 16 and after the search export PR has been merged (this PR is branched off the **feature/search-export** branch).

### Risks
- None foreseen.

### Follow-up actions
- This PR adds the pyparsing Python library, so this should be installed when deployed.

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
- [ ] **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
- [ ] **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?** 
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: cadasta/search/parser.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            # Fuzzy match clause
+            if has_fuzziness and len(term) > 1:
+                dsl.append({
+                    'query': term,
+                    'fields': fields,
+                    'fuzziness': get_fuzziness(term),
+                    'prefix_length': 1,
+                })
+
+    return [{'multi_match': x} for x in dsl]
```
Comment (oliverroick) [inline cadasta/search/parser.py]: Is there a way to create the list that is returned while looping through `terms`? It seems overhead to have two loops here.

File: cadasta/templates/search/search.html [hunk 2]
```diff
@@ -31,9 +31,12 @@
                 <div id="search-guidelines" class="hidden small">
                   {% blocktrans %}
                   <ul>
-                    <li>Review your wording, make sure that there aren’t any typos in your search query</li>
-                    <li>If no results match your query, try removing some search terms (<kbd>John</kbd>, instead of <kbd>John White</kbd>)</li>
-                    <li>Search results are limited to a maximum of {{ max_num_results }} entries</li>
+                    <li>By default, search results are required to only match at least one of the provided terms.
```
Comment (seav) [inline cadasta/templates/search/search.html]: Okay, I've removed "only".

File: cadasta/search/parser.py (hunk 1 above)
Comment (seav) [inline cadasta/search/parser.py]: Yes, there is a way. I've recoded the function to just have one loop.

**Comment by seav** [comment]
> "I'm assuming the other commits are from the search export PR?!"

That's correct. I've branched this PR off from the search export PR branch. So this PR is also showing the search export commits because they're not present in the master branch.

**Pull request merged**